platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/extensions/property/PropertyUpgradeAnnotatedMethodReader.java

```diff
@@ -117,21 +124,94 @@ public Collection<Result> readRequest(ExecutableElement method) {
+                        break;
+                    case SETTER:
+                        CallInterceptionRequest jvmSetterRequest = createJvmSetterInterceptionRequest(accessorSpec, method);
+                        requests.add(jvmSetterRequest);
+                        break;
+                    default:
+                        throw new IllegalArgumentException("Unsupported accessor type: " + accessorSpec.accessorType);
+                }
+            }
+            return requests.stream()
+                .map(Success::new)
+                .collect(Collectors.toList());
         } catch (AnnotationReadFailure failure) {
             return Collections.singletonList(new InvalidRequest(failure.reason));
         }
     }
 
+    static boolean isProviderOrProperty(Type type) {
+        return type.equals(PROVIDER_TYPE) || type.equals(PROPERTY_TYPE);
+    }
+
+    @SuppressWarnings("unchecked")
+    private List<AccessorSpec> readAccessorSpecsFromUpgradedProperty(ExecutableElement method, AnnotationMirror annotationMirror) {
+        List<AnnotationMirror> originalAccessors = AnnotationUtils.findAnnotationValue(annotationMirror, "originalAccessors")
+            .map(v -> (List<AnnotationMirror>) v.getValue())
+            .orElse(Collections.emptyList());
+        if (!originalAccessors.isEmpty()) {
+            return originalAccessors.stream()
+                .map(annotation -> getAccessorSpec(method, annotation))
+                .collect(Collectors.toList());
+        }
+        return Arrays.asList(
+            getAccessorSpec(method, AccessorType.GETTER, annotationMirror),
+            getAccessorSpec(method, AccessorType.SETTER, annotationMirror)
+        );
+    }
+
+    private AccessorSpec getAccessorSpec(ExecutableElement method, AnnotationMirror annotation) {
+        String methodName = AnnotationUtils.findAnnotationValue(annotation, "methodName")
+            .map(v -> (String) v.getValue())
+            .orElseThrow(() -> new AnnotationReadFailure("Missing 'methodName' attribute in @UpgradedAccessor"));
+        AccessorType accessorType = AnnotationUtils.findAnnotationValue(annotation, "value")
+            .map(v -> AccessorType.valueOf(v.getValue().toString()))
+            .orElseThrow(() -> new AnnotationReadFailure("Missing 'value' attribute in @UpgradedAccessor"));
+        Type originalType = extractOriginalType(method, annotation);
+        return getAccessorSpec(accessorType, methodName, originalType, annotation);
+    }
+
+    private AccessorSpec getAccessorSpec(ExecutableElement method, AccessorType accessorType, AnnotationMirror annotation) {
+        String propertyName = getPropertyName(method);
+        Type originalType = extractOriginalType(method, annotation);
+        String methodName;
+        switch (accessorType) {
+            case GETTER:
+                String capitalize = propertyName.substring(0, 1).toUpperCase() + propertyName.substring(1);
```

Comment:
❓ Do we have a util class for this property name <-> getter/setter logic already? If not, we really should have. Maybe add it to `PropertyAccessorType`?